File: subtitler/subtitle_parser.py

from typing import Final
from enum import Enum
from subtitle_entities import SubtitleFile, SubtitleEntry
from print_utils import verbose
import re

class SubtitleParser:

    INDEX_PATTERN:     Final[re.Pattern] = re.compile(r"^\d+$")
    TIMESTAMP_PATTERN: Final[re.Pattern] = re.compile(r"^[\d:,]+\s+\-\-\>\s+[\d:,]+$")

    class Type(Enum):
        INDEX, TIMESTAMP, CONTENT = range(3)

    entries: list[SubtitleEntry] = []

    cur_index = ""
    cur_timestamp = ""
    cur_content = []
    last_parsed: Type

    def parse(self, file_path: str) -> SubtitleFile:
        
        with open(file_path) as filehandle:
            lines = filehandle.readlines()
        
        num_lines = len(lines)
        verbose("Number of lines: " + str(num_lines))

        for i, line in enumerate(lines):
            line = line.strip()
            if not line:
                continue
            
            if i < num_lines-1 and self.looks_like_index(line) and self.looks_like_timestamp(lines[i+1]):
                self.reset()
                self.cur_index = self.INDEX_PATTERN.match(line).group()
                self.last_parsed = self.Type.INDEX
                continue

            if self.cur_index and self.looks_like_timestamp(line):
                self.cur_timestamp = self.TIMESTAMP_PATTERN.match(line).group()
                self.last_parsed = self.Type.TIMESTAMP
                continue

            if self.last_parsed in [self.Type.TIMESTAMP, self.Type.CONTENT]:
                self.cur_content.append(line)
                self.last_parsed = self.Type.CONTENT
                continue
        
        self.reset()
        
        return SubtitleFile(file_path, self.entries)


    def reset(self) -> None:
        if self.cur_index and self.cur_timestamp:
            entry = SubtitleEntry(self.cur_index, self.cur_timestamp, self.cur_content)
            verbose("\nParsed entry: \n" + str(entry))
            self.entries.append(entry)

        self.last_parsed = None
        self.cur_index = ""
        self.cur_timestamp = ""
        self.cur_content = []
    
    def looks_like_index(self, line: str) -> bool:
        line = line.strip()
        return self.__matches(self.INDEX_PATTERN, line)
    
    def looks_like_timestamp(self, line: str) -> bool:
        line = line.strip()
        return self.__matches(self.TIMESTAMP_PATTERN, line)

    @staticmethod
    def __matches(pattern: re.Pattern, line: str) -> bool:
        if pattern.match(line):
            return True
        else:
            return False

File: src/main/java/se/p950tes/subtitler/tooling/parsing/SubtitleParser.java
package se.p950tes.subtitler.tooling.parsing;

import java.io.BufferedReader;
import java.io.IOException;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;
import java.util.regex.Pattern;

import se.p950tes.subtitler.io.FileManager;
import se.p950tes.subtitler.logging.Logger;
import se.p950tes.subtitler.model.SubtitleEntry;
import se.p950tes.subtitler.model.SubtitleFile;

public class SubtitleParser {

	private static final Pattern INDEX_PATTERN = Pattern.compile("^\\d+$");
	private static final Pattern TIMESTAMP_PATTERN = Pattern.compile("^[\\d:,]+\\s+" + Pattern.quote("-->") + "\\s+[\\d:,]+$");

	private static enum Type {
		INDEX, TIMESTAMP, CONTENT;
	}

	private final List<SubtitleEntry> entries = new ArrayList<>();
	private final FileManager fileManager;
	private final Logger logger;
	private final Path file;

	
	private BufferedReader fileReader;
	private String currentIndex;
	private String currentTimestamp;
	private List<String> currentContent = new ArrayList<>();
	private Type lastParsed;

	public SubtitleParser(Logger logger, FileManager fileManager, Path file) {
		this.fileManager = fileManager;
		this.logger = logger;
		this.file = file;
	}

	public SubtitleFile parse() {
		logger.verbose("Parsing: " + file);
		this.fileReader = fileManager.getFileReader(file);
		
		String currentLine = readNextLine();
		while (currentLine != null) {
			String nextLine = readNextLine();
			
			if (looksLikeIndex(currentLine) && looksLikeTimestamp(nextLine)) {
				saveCurrentAndStartNewEntry();
				currentIndex = currentLine;
				lastParsed = Type.INDEX;
			
			} else if (currentIndex != null && looksLikeTimestamp(currentLine)) {
				currentTimestamp = currentLine;
				lastParsed = Type.TIMESTAMP;
				
			} else if (lastParsed == Type.TIMESTAMP || lastParsed == Type.CONTENT) {
				currentContent.add(currentLine);
				lastParsed = Type.CONTENT;
			}
			currentLine = nextLine;
		}
		saveCurrentAndStartNewEntry();
		logger.verbose("Entries found: " + entries.size());
		return new SubtitleFile(file, entries);
	}
	
	private String readNextLine() {
		try {
			String nextLine = "";
			while (nextLine.isEmpty()) {
				nextLine = fileReader.readLine();
				if (nextLine == null) {
					// End of file
					return null;
				}
				nextLine = nextLine.trim();
			}
			return nextLine;
		} catch (IOException e) {
			throw new RuntimeException("Failed to read line from file", e);
		}
	}

	private void saveCurrentAndStartNewEntry() {
		if (currentIndex != null && currentTimestamp != null) {
			var entry = new SubtitleEntry(Integer.parseInt(currentIndex), currentTimestamp, currentContent);
			entries.add(entry);
		}
		lastParsed = null;
		currentIndex = null;
		currentTimestamp = null;
		currentContent = new ArrayList<>();
	}
	private static boolean looksLikeIndex(String line) {
		if (line == null) {
			return false;
		}
		return INDEX_PATTERN.matcher(line).matches();
	}
	private static boolean looksLikeTimestamp(String line) {
		if (line == null) {
			return false;
		}
		return TIMESTAMP_PATTERN.matcher(line).matches();
	}
}
